compositor_pipeline/src/pipeline/whip_whep.rs

```diff
@@ -0,0 +1,200 @@
+};
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    sync::{Arc, Mutex, Weak},
+    time::Instant,
+};
+use tokio::sync::mpsc;
+use tracing::{info, warn};
+use webrtc::{
+    api::{
+        interceptor_registry::register_default_interceptors,
+        media_engine::{MediaEngine, MIME_TYPE_H264, MIME_TYPE_OPUS},
+        APIBuilder,
+    },
+    ice_transport::ice_server::RTCIceServer,
+    interceptor::registry::Registry,
+    peer_connection::{configuration::RTCConfiguration, RTCPeerConnection},
+    rtp_transceiver::{
+        rtp_codec::{RTCRtpCodecCapability, RTCRtpCodecParameters, RTPCodecType},
+        rtp_transceiver_direction::RTCRtpTransceiverDirection,
+        RTCRtpTransceiverInit,
+    },
+};
+
+mod config;
+mod handlers;
+mod utils;
+
+use tokio::task;
+
+use crate::{queue::PipelineEvent, Pipeline};
+
+use super::EncodedChunk;
+
+pub(crate) const VIDEO_PAYLOAD_TYPE: u8 = 96;
+pub(crate) const AUDIO_PAYLOAD_TYPE: u8 = 111;
+
+#[tokio::main]
+pub async fn run_whip_whep_server(pipeline: Weak<Mutex<Pipeline>>) {
+    let config = read_config();
+    let port = config.api_port;
+
+    if !config.start_whip_whep {
+        return;
+    }
+
+    let pipeline_ctx = match pipeline.upgrade() {
+        Some(pipeline) => pipeline.lock().unwrap().ctx.clone(),
+        None => {
+            warn!("Pipeline stopped.");
+            return;
+        }
+    };
+
+    let state = pipeline_ctx.whip_whep_state.clone();
+
+    let app = Router::new()
+        .route("/status", get(status))
+        .route("/whip/:id", post(handle_whip))
+        .route("/whip/:id", options(handle_options))
+        .route("/session/:id", patch(whip_ice_candidates_handler))
+        .route("/session/:id", delete(terminate_whip_session))
+        .with_state(state);
+
+    let listener = tokio::net::TcpListener::bind(SocketAddr::from(([0, 0, 0, 0], port)))
+        .await
+        .unwrap();
+
+    let server_task = task::spawn(async {
+        let _ = axum::serve(listener, app).await;
+    });
+    info!("started http server");
+    if let Err(e) = server_task.await {
+        eprintln!("WHIP/WHEP server task failed: {:?}", e);
+    }
+}
+
+#[derive(Debug)]
+pub struct InputConnectionUtils {
+    pub video_sender: Option<mpsc::Sender<PipelineEvent<EncodedChunk>>>,
+    pub audio_sender: Option<mpsc::Sender<PipelineEvent<EncodedChunk>>>,
+    pub bearer_token: Option<String>,
+    pub peer_connection: Option<Arc<RTCPeerConnection>>,
+    pub start_time_vid: Option<Instant>,
+    pub start_time_aud: Option<Instant>,
+    pub depayloader: Arc<Mutex<Depayloader>>,
+}
+
+#[derive(Debug)]
+pub struct WhipWhepState {
+    pub input_connections: Arc<Mutex<HashMap<InputId, InputConnectionUtils>>>,
+}
+
+impl WhipWhepState {
+    pub fn new() -> Arc<Self> {
+        Arc::new(WhipWhepState {
+            input_connections: Arc::from(Mutex::new(HashMap::new())),
+        })
+    }
+}
+
+pub async fn init_pc(
```

Comment:
please look through my comments on whip client PR, some of the stuff from there applies in this PR
e.g. init_peer_connection